File: MusicList/app/src/main/java/com/example/musiclist/adapter/MusicAdapter.java
package com.example.musiclist.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.example.musiclist.R;
import com.example.musiclist.model.MusicModel;
import com.example.musiclist.model.MusicResult;
import com.example.musiclist.util.IItemClickListener;

import java.util.List;

public class MusicAdapter extends RecyclerView.Adapter<MusicAdapter.ViewHolder> {

    private List<MusicModel> musicList;
    private int rowLayout;
    private Context mContext;
    public String price;

    public MusicAdapter(MusicResult musicList, int rowLayout, Context mContext) {
        this.musicList = musicList.getResults();
        this.rowLayout = rowLayout;
        this.mContext = mContext;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(rowLayout, parent, false);
        return new ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        final MusicModel mm = musicList.get(position);

        price = mm.getTrackPrice() <= 0 ? "Free" : Double.toString(mm.getTrackPrice());

        holder.trackName.setText(mm.getTrackName());
        holder.artistName.setText(mm.getArtistName());
        holder.trackPrice.setText(price);

        Glide.with(mContext)
                .load(mm.getArtworkUrl100())
                .into(holder.artworkImage100);
        holder.setClickListener((view, i, isLongClick) -> {
            if (isLongClick) {
                Toast.makeText(mContext, "#" + i + " - " + mm.getTrackName() + " (Long click)", Toast.LENGTH_SHORT).show();
            } else {
                Toast.makeText(mContext, "#" + i + " - " + mm.getTrackName(), Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    public int getItemCount() {
        return musicList == null ? 0 : musicList.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener, View.OnLongClickListener {

        public TextView trackName;
        public TextView artistName;
        public TextView trackPrice;
        public ImageView artworkImage100;

        private IItemClickListener clickListener;

        public ViewHolder(View itemView) {
            super(itemView);

            trackName = (TextView) itemView.findViewById(R.id.txtTrackName);
            artistName = (TextView) itemView.findViewById(R.id.txtArtName);
            trackPrice = (TextView) itemView.findViewById(R.id.txtTrackPrice);
            artworkImage100 = (ImageView) itemView.findViewById(R.id.artistImg100);

            itemView.setTag(itemView);
            itemView.setOnClickListener(this);
            itemView.setOnLongClickListener(this);
        }

        public void setClickListener(IItemClickListener itemClickListener) {
            this.clickListener = itemClickListener;
        }

        @Override
        public void onClick(View v) {
            clickListener.onClick(v, getPosition(), false);
        }

        @Override
        public boolean onLongClick(View v) {
            clickListener.onClick(v, getPosition(), true);
            return false;
        }
    }
}

File: KotlinMusic/app/src/main/java/com/example/kotlinmusic/adapter/MusicAdapter.kt
package com.example.kotlinmusic.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.example.kotlinmusic.model.MusicModel
import com.example.kotlinmusic.model.MusicResult
import com.example.kotlinmusic.util.IItemClickListener
import kotlinx.android.synthetic.main.music_row.view.*


class MusicAdapter(
    musicResult: MusicResult,
    private val rowLayout: Int,
    private var mContext: Context
): RecyclerView.Adapter<MusicAdapter.ViewHolder>() {

    private var musicList: List<MusicModel> = musicResult.results

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val v =
            LayoutInflater.from(parent.context).inflate(rowLayout, parent, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val mm: MusicModel = musicList[position]

        if (mm.trackName.isNullOrBlank()) holder.trackName.text = "Unavailable"
        else holder.trackName.text = mm.trackName

        if (mm.trackPrice!! <= 0) holder.trackPrice.text = "Free"
        else holder.trackPrice.text = mm.trackPrice.toString()

        holder.artistName.text = mm.artistName

        Glide.with(mContext)
            .load(mm.artworkUrl100)
            .into(holder.artworkImage100)
        holder.setClickListener(object : IItemClickListener {
            override fun onClick(view: View?, position: Int, isLongClick: Boolean) {
                if (isLongClick) {
                    Toast.makeText(
                        mContext,
                        "#" + position + " - " + mm.trackName + " (Long click)",
                        Toast.LENGTH_SHORT
                    ).show()
                } else {
                    Toast.makeText(
                        mContext,
                        "#" + position + " - " + mm.trackName,
                        Toast.LENGTH_SHORT
                    )
                        .show()
                }
            }
        })
    }

    override fun getItemCount(): Int {
        return if (musicList == null) 0 else musicList.size
    }


    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener, View.OnLongClickListener {

        init {
            itemView.tag = itemView
            itemView.setOnClickListener(this)
            itemView.setOnLongClickListener(this)
        }

        val trackName: TextView = itemView.txtTrackName
        val artistName: TextView = itemView.txtArtName
        val trackPrice: TextView = itemView.txtTrackPrice
        val artworkImage100: ImageView = itemView.artistImg100
        private var clickListener: IItemClickListener? = null

        fun setClickListener(itemClickListener: IItemClickListener?) {
            clickListener = itemClickListener
        }

        override fun onClick(v: View?) {
            clickListener!!.onClick(v, position, false)
        }

        override fun onLongClick(v: View?): Boolean {
            clickListener!!.onClick(v, position, true)
            return false
        }
    }
}